File: 백준/Silver/2468. 안전 영역/안전 영역.swift
let n = Int(readLine()!)!

var arr: [[Int]] = Array(repeating: Array(repeating: 0, count: n), count: n)
var visited: [[Bool]] = [[]]
var min = 101
var max = 0

let dx = [0,0,1,-1]
let dy = [1,-1,0,0]

for i in 0...n-1{
    let input = readLine()!.split(separator: " ").map{Int(String($0))!}
    for j in 0...n-1{
        arr[i][j] = input[j]
        if(max < arr[i][j]){
            max = arr[i][j]
        }
        if(min > arr[i][j]){
            min = arr[i][j]
        }
    }
}
var ans:[Int] = []
for k in min...max{
    visited = Array(repeating: Array(repeating: false, count: n), count: n)
    var cnt = 0
    for i in 0...n-1{
        for j in 0...n-1{
            if(arr[i][j] <= k){
                visited[i][j] = true
            }
        }
    }

    for i in 0...n-1{
        for j in 0...n-1{
            if(!visited[i][j]){
                dfs(i,j)
                cnt += 1
            }
        }
    }
    ans.append(cnt)
}
var aa = ans.sorted()

if(aa.last! == 0){
    print(1)
}
else{
    print(aa.last!)
}

func dfs(_ x:Int,_ y:Int){
    visited[x][y] = true
    for k in 0...3{
        var dx = x + dx[k]
        var dy = y + dy[k]
        if(dx >= 0 && dx < n && dy >= 0 && dy < n){
            if(!visited[dx][dy]){
                dfs(dx, dy)
            }
        }
    }
}

File: 백준/Silver/2468. 안전 영역/안전 영역.java
import java.io.*;
import java.util.*;

public class Main {

    static int N;
    static int [] dx = {-1,1,0,0};
    static int [] dy = {0,0,1,-1};
    static int [][] arr;
    static boolean [][] check;
    static int max = 0;
    static int cnt = 0;

    static void dfs(int x,int y){
        check[x][y] = true;

        for (int i = 0; i < 4; i++) {
            int ax = x + dx[i];
            int ay = y + dy[i];

            if(ax>=0 && ay>=0 && ax < N && ay < N && !check[ax][ay]){
                dfs(ax,ay);
            }
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        int ans = 1;
        N = Integer.parseInt(br.readLine());
        arr = new int[N][N];
        for (int i = 0; i < N; i++) {
            String [] A = br.readLine().split(" ");
            for (int j = 0; j < N; j++) {
                arr[i][j] = Integer.parseInt(A[j]);
                if(max < arr[i][j]){
                    max = arr[i][j];
                }
            }
        }

        for (int k = 1; k <= max; k++) {
            check = new boolean[N][N];
            cnt = 0;
            for (int i = 0; i < N; i++) {
                for (int j = 0; j < N; j++) {
                    if(arr[i][j] <= k){
                        check[i][j] = true;
                    }
                }
            }

            for (int i = 0; i < N; i++) {
                for (int j = 0; j < N; j++) {
                    if(!check[i][j]){
                        dfs(i,j);
                        cnt++;
                    }
                }
            }
            //System.out.println(cnt);
            if(cnt > ans){
                ans = cnt;
            }

        }

        System.out.println(ans);
    }


}
